File: LeetCode/53.Maximum_Subarray/53.Maximum_Subarray.py
#!/usr/bin/env python3


# dp
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        n = len(nums)
        # define dp[i] include current ith num max sum val

        # init dp
        dp = [0] * n
        dp[0] = nums[0]

        res = dp[0]

        for i in range(1, n):
            dp[i] = max(dp[i - 1] + nums[i], nums[i])
            res = max(res, dp[i])

        return res


# Kadane's algo
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        res, cur = -math.inf, 0
        for i, num in enumerate(nums):
            cur += num
            res = max(cur, res)
            if cur < 0:
                cur = 0

        return res

File: LeetCode/53.Maximum_Subarray/53.Maximum_Subarray.swift
class Solution {
    func maxSubArray(_ nums: [Int]) -> Int {
        var res = Int.min
        var sub = 0

        for num in nums {
            sub += num
            res = max(res, sub)
            if sub < 0 {
                sub = 0
            }
        }

        return res
    }
}

class Solution {
    /// 返回最大连续子数组和
    func maxSubArray(_ nums: [Int]) -> Int {
        // 题目保证至少一个元素；若要健壮化可额外判断空数组
        var cur = nums[0] // 以当前索引结尾的最大和
        var ans = nums[0] // 全局最大和

        for i in 1 ..< nums.count {
            // 若 cur 为负，会拖累后续，直接从 nums[i] 重新开始
            cur = max(nums[i], cur + nums[i])
            ans = max(ans, cur)
        }
        return ans
    }
}
